databroker/core.py

```diff
@@ -0,0 +1,543 @@
+        mapping spec names (strings) to handlers (callable classes)
+    handler_override : callable class, optional
+        overrides registered handlers
+        
+
+    Example
+    -------
+    >>> header = DataBroker[-1]
+    >>> images = Images(header, 'my_detector_lightfield')
+    >>> for image in images:
+            # do something
+    """
+    return Images(mds, fs, headers, name, handler_registry, handler_override)
+
+
+class Images(FramesSequence):
+    def __init__(self, mds, fs, headers, name, handler_registry=None,
+                 handler_override=None):
+        """
+        Load images from a detector for given Header(s).
+
+        Parameters
+        ----------
+        fs : FileStoreRO
+        headers : Header or list of Headers
+        name : str
+            field name (data key) of a detector
+        handler_registry : dict, optional
+            mapping spec names (strings) to handlers (callable classes)
+        handler_override : callable class, optional
+            overrides registered handlers
+
+        Example
+        -------
+        >>> header = DataBroker[-1]
+        >>> images = Images(header, 'my_detector_lightfield')
+        >>> for image in images:
+                # do something
+        """
+        self.fs = fs
+        events = get_events(mds, fs, headers, [name], fill=False)
+        self._datum_uids = [event.data[name] for event in events
+                            if name in event.data]
+        self._len = len(self._datum_uids)
+        first_uid = self._datum_uids[0]
+        if handler_override is None:
+            self.handler_registry = handler_registry
+        else:
+            # mock a handler registry
+            self.handler_registry = defaultdict(lambda: handler_override)
+        with self.fs.handler_context(self.handler_registry) as fs:
+            example_frame = fs.get_datum(first_uid)
+        # Try to duck-type as a numpy array, but fall back as a general
+        # Python object.
+        try:
+            self._dtype = example_frame.dtype
+        except AttributeError:
+            self._dtype = type(example_frame)
+        try:
+            self._shape = example_frame.shape
+        except AttributeError:
+            self._shape = None  # as in, unknown
+
+    @property
+    def pixel_type(self):
+        return self._dtype
+
+    @property
+    def frame_shape(self):
+        return self._shape
+
+    def __len__(self):
+        return self._len
+
+    def get_frame(self, i):
+        with self.fs.handler_context(self.handler_registry) as fs:
+            img = fs.get_datum(self._datum_uids[i])
+        if hasattr(img, '__array__'):
+            return Frame(img, frame_no=i)
+        else:
+            # some non-numpy-like type
+            return img
+
+
+def _inspect_descriptor(descriptor):
+    """
+    Return a dict with the data keys mapped to boolean answering whether
+    data is external.
+    """
+    # TODO memoize to cache these results
```

Comment:
@tacaswell already wrote code to do this (using the boltons cache) which I have since cherry-picked. No harm since the work is already done!